### PR TITLE
Global notification when adding/removing projects

### PR DESCRIPTION
**Pull request by rchande**
Adding or removing a project fires a WorkspaceChanged event for each
file in the project, which starts more analysis from the Solution
Crawler. This massively slows down project load/reload. To improve this,
start a global notification of project load/reload, which will cause
Solution Crawler to only run active file analysis.Project unload is about 10 
seconds faster with this change.

### CODE REVIEW

**Comment by rchande** [comment]
@heejaechang Want to take look? I'll wait until after the RI to merge.

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs [hunk 1]
```diff
@@ -539,20 +540,24 @@ private CompilationTracker GetCompilationTracker(ProjectId projectId)
 
         private Solution AddProject(ProjectId projectId, ProjectState projectState)
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs]: does this already contain all files/references added? I remember we did every in sequence before. like add a project, and a file one by one, add a reference one by one. and change project configuration one by one and etc.

did we change that? otherwise, the scope of global operation seems too small

Comment (heejaechang) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs]: also, global operation can be nested. so, if there is a point for solution re/loading (like KnownUIContext - http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/SolutionBuildMonitor.cs,cd3fac6641d35e3e), you can start global operation there as well without worrying about individual add/remove project case in already loaded solution.

Comment (rchande) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs]: Good catch--I originally had this in ForkProject and intended to move it into AddDocument and RemoveDocument.

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs [hunk 2]
```diff
@@ -539,20 +540,24 @@ private CompilationTracker GetCompilationTracker(ProjectId projectId)
...
         {
-            var newProjectIds = _projectIds.ToImmutableArray().Add(projectId);
-            var newStateMap = _projectIdToProjectStateMap.Add(projectId, projectState);
-            var newDependencyGraph = CreateDependencyGraph(newProjectIds, newStateMap);
-            var newTrackerMap = CreateCompilationTrackerMap(projectId, newDependencyGraph);
-            var newLinkedFilesMap = CreateLinkedFilesMapWithAddedProject(newStateMap[projectId]);
+            var notificationService = this.Workspace.Services.GetService<IGlobalOperationNotificationService>();
+            using (notificationService?.Start("Remove Project"))
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs]: Should be "Add Project"?

**Comment by Pilchie** [comment]
LGTM

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs [hunk 3]
```diff
@@ -1119,17 +1120,21 @@ public Solution WithProjectAnalyzerReferences(ProjectId projectId, IEnumerable<A
 
         private Solution AddDocument(DocumentState state)
         {
-            if (state == null)
+            var notificationService = this.Workspace.Services.GetService<IGlobalOperationNotificationService>();
+            using (notificationService?.Start("Remove Document"))
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs]: Should this be "Add Document"?

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs [hunk 4]
```diff
@@ -1119,17 +1120,21 @@ public Solution WithProjectAnalyzerReferences(ProjectId projectId, IEnumerable<A
 
         private Solution AddDocument(DocumentState state)
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs]: Is there no way to pull this out to the larger project load/unload, rather than doing this a bunch of times?

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs (hunk 3 above)
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs]: How does this actually work? Does it tell the solution crawler to permanently ignore any workspace change events during this time? If so, then at the end of this process how does solution crawler know to go re-analyze everything now that we're done (there will be no workspace change event to trigger it). Or is it just a batcher that gathers up the events during global operations and then handles them all at once in the end? If so, then what's to stop it from triggering between each of these mini-batches (unless we move it up to the project-level load/unload somehow).

**Comment by dpoeschl** [comment]
There are three commits with the same title. Do you intend to squash them?

File: src/VisualStudio/Core/Def/SolutionBuildMonitor.cs [hunk 5]
```diff
@@ -42,6 +43,16 @@ public void Dispose()
 
         private void SolutionBuildingContextChanged(object sender, UIContextChangedEventArgs e)
         {
+            ContextChangedWorker(e, "Solution Building");
+        }
+
+        private void SolutionOpeningContext_UIContextChanged(object sender, UIContextChangedEventArgs e)
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Def/SolutionBuildMonitor.cs]: How is this change related?

File: src/VisualStudio/Core/Def/SolutionBuildMonitor.cs [hunk 6]
```diff
@@ -22,6 +22,7 @@ public SolutionBuildMonitor(VisualStudioWorkspace workspace)
             {
                 _notificationService = notificationService;
                 KnownUIContexts.SolutionBuildingContext.UIContextChanged += SolutionBuildingContextChanged;
+                KnownUIContexts.SolutionOpeningContext.UIContextChanged += SolutionOpeningContext_UIContextChanged;
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/SolutionBuildMonitor.cs]: you probably want to change name of this file. to like SolutionContextMonitoring.

File: src/VisualStudio/Core/Def/SolutionBuildMonitor.cs [hunk 7]
```diff
@@ -51,7 +62,7 @@ private void SolutionBuildingContextChanged(object sender, UIContextChangedEvent
                         _operation.Dispose();
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/SolutionBuildMonitor.cs]: operation can be interleaved. you should put operation in a map or something with operation as a key.

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs (hunk 3 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs]: solution crawler always batch. globaloperation pause processor (that actually run code), but queue is still alive. so, when solution crawler get resumed, it will process pending work in the queue in priority order based on current context (what is active file, what is opened file and etc)

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs [hunk 8]
```diff
@@ -1119,17 +1120,21 @@ public Solution WithProjectAnalyzerReferences(ProjectId projectId, IEnumerable<A
...
             {
-                throw new ArgumentNullException(nameof(state));
-            }
+                if (state == null)
+                {
+                    throw new ArgumentNullException(nameof(state));
+                }
 
-            CheckContainsProject(state.Id.ProjectId);
+                CheckContainsProject(state.Id.ProjectId);
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs]: does this need to be in workspace layer? can VS layer have this global operation? unlike workspace layer where we do things one by one. VS layer might have better way to scope it. doesn't it?

Comment (rchande) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs]: It looks like the project system calls AbsractProject.Disconnect() when a project is unloaded. We then start removing documents ourselves, so we can just start the global operation first.

When the project is reloaded, we see the file additions one at a time, so I'm not sure how to batch those.

Comment (heejaechang) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs]: hmmm. there should be some project load event from vs. let me take a quick look.

Comment (heejaechang) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs]: http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/VisualStudioProjectTracker_IVsSolutionEvents.cs,c60564eef81e7810

**Comment by mattwar** [comment]
First of all what is this global notification service, and why is it in the workspace layer, and why of all places is it being sent during AddDocument on a Solution object?

**Comment by heejaechang** [comment]
@mattwar 
Global operation notification service - probably bad name - is something I added long time ago. basically any blocking long running operation can raise global operation notification, which will notify listeners so that they can pause what they were doing until global operation is finished.

currently I believe only listener is solution crawler. and the global notification is raised by "Wait Indicator - the model dialogbox VS shows when UI thread is busy more than 2 seconds" and "Build".

Ravi is trying to add it to solution/project load. but probably we need to move it up to VS layer.

**Comment by rchande** [comment]
@mattwar I've moved the notifications from Solution up into the VS Workspace impl.

**Comment by mattwar** [comment]
:+1:

File: src/VisualStudio/Core/Def/SolutionEventMonitor.cs [hunk 9]
```diff
@@ -1,63 +1,72 @@
...
     /// </summary>
-    internal sealed class SolutionBuildMonitor : IDisposable
+    internal sealed class SolutionEventMonitor : IDisposable
     {
         private GlobalOperationNotificationService _notificationService;
-        private GlobalOperationRegistration _operation;
-
-        public SolutionBuildMonitor(VisualStudioWorkspace workspace)
+        private Dictionary<string, GlobalOperationRegistration> _operations = new Dictionary<string, GlobalOperationRegistration>();
+        public SolutionEventMonitor(VisualStudioWorkspace workspace)
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/SolutionEventMonitor.cs]: new line?

File: src/VisualStudio/Core/Def/SolutionEventMonitor.cs [hunk 10]
```diff
@@ -1,63 +1,72 @@
...
-                _operation = null;
+                globalOperation.Dispose();
             }
 
+            _operations.Clear();
+
             if (_notificationService != null)
             {
                 _notificationService = null;
                 KnownUIContexts.SolutionBuildingContext.UIContextChanged -= SolutionBuildingContextChanged;
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/SolutionEventMonitor.cs]: shouldn't you unsubscribe your event too?

File: src/VisualStudio/Core/Def/SolutionEventMonitor.cs [hunk 11]
```diff
@@ -1,63 +1,72 @@
...
         private void SolutionBuildingContextChanged(object sender, UIContextChangedEventArgs e)
         {
+            ContextChangedWorker(e, "Solution Building");
+        }
+
+        private void SolutionOpeningContext_UIContextChanged(object sender, UIContextChangedEventArgs e)
+        {
+            ContextChangedWorker(e, "Solution Opening");
+        }
+        private void ContextChangedWorker(UIContextChangedEventArgs e, string operation)
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/SolutionEventMonitor.cs]: new line?

File: src/VisualStudio/Core/Def/SolutionEventMonitor.cs [hunk 12]
```diff
@@ -1,63 +1,72 @@
...
         }
-
         private void SolutionBuildingContextChanged(object sender, UIContextChangedEventArgs e)
         {
+            ContextChangedWorker(e, "Solution Building");
+        }
+
+        private void SolutionOpeningContext_UIContextChanged(object sender, UIContextChangedEventArgs e)
+        {
+            ContextChangedWorker(e, "Solution Opening");
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/SolutionEventMonitor.cs]: any UI context change? interesting.

**Comment by heejaechang** [comment]
:+1:

File: src/VisualStudio/Core/Def/SolutionEventMonitor.cs (hunk 9 above)
Comment (rchande) [inline src/VisualStudio/Core/Def/SolutionEventMonitor.cs]: Managed to use git in a way that lost my newlines, fixed.

File: src/VisualStudio/Core/Def/SolutionEventMonitor.cs (hunk 10 above)
Comment (rchande) [inline src/VisualStudio/Core/Def/SolutionEventMonitor.cs]: Fixed.
